metricflow/protocols/sql_request.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import annotations
+
+import typing
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import Enum
+from operator import itemgetter
+from typing import Optional, Sequence, Dict
+
+import pandas as pd
+from pydantic import Field
+
+from metricflow.model.objects.base import FrozenBaseModel
+from metricflow.object_utils import assert_exactly_one_arg_set
+
+
+@dataclass(frozen=True)
+class SqlRequestId:
+    """Identifies a request (i.e. a call to SqlClient.query() or SqlClient.execute()) to the SQL engine."""
+
+    id_str: str
+
+    def __repr__(self) -> str:  # noqa: D
+        return self.id_str
+
+
+@dataclass(frozen=True)
+class SqlRequestResult:
+    """The result from a request to the SQL engine."""
+
+    df: Optional[pd.DataFrame] = None
+    exception: Optional[Exception] = None
+
+    def __post_init__(self) -> None:  # noqa: D
+        assert_exactly_one_arg_set(df=self.df is not None, exception=self.exception)
```

Comment:
Is this boolean conversion because an empty dataframe is falsey? 

Also, that's an unexpected bit of news for me about assert_exactly_one_arg_set - I assumed it was doing `is None` checks. So does its docstring, so I just put up #290

Reply:
Updated, taking #290 into account.